lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -926,14 +926,62 @@ class ADContext {
     return dfi;
   }
 
+public:
+  template <typename ...T, typename ...U>
+  InFlightDiagnostic diagnose(SourceLoc loc, Diag<T...> diag,
+                              U &&...args) const {
+    return getASTContext().Diags.diagnose(loc, diag, std::forward<U>(args)...);
+  }
+
+  /// Given an instruction and a differentiation task associated with the
+  /// parent function, emits a "not differentiable" error based on the task. If
+  /// the task is indirect, emits notes all the way up to the outermost task,
+  /// and emits an error at the outer task. Otherwise, emits an error directly.
+  template<typename ...T, typename ...U>
+  InFlightDiagnostic emitNondifferentiabilityError(
+      SILInstruction *inst, DifferentiationInvoker invoker,
+      Diag<T...> diag, U &&...args);
+
+  /// Given a value and a differentiation task associated with the parent
+  /// function, emits a "not differentiable" error based on the task. If the
+  /// task is indirect, emits notes all the way up to the outermost task, and
+  /// emits an error at the outer task. Otherwise, emits an error directly.
+  template<typename ...T, typename ...U>
+  InFlightDiagnostic emitNondifferentiabilityError(
+      SILValue value, DifferentiationInvoker invoker,
+      Diag<T...> diag, U &&...args);
+
+  /// Emit a "not differentiable" error based on the given differentiation task
+  /// and diagnostic.
+  template<typename ...T, typename ...U>
+  InFlightDiagnostic emitNondifferentiabilityError(
+      SourceLoc loc, DifferentiationInvoker invoker,
+      Diag<T...> diag, U &&...args);
+};
+
+class DifferentiationTransformer {
 private:
+  /// Reference to the main transform.
+  SILModuleTransform &transform;
+
+  /// Context necessary for performing the transformations.
+  ADContext context;
+
   /// Promotes the given `differentiable_function` instruction to a valid
   /// `@differentiable` function-typed value.
   SILValue promoteToDifferentiableFunction(
       DifferentiableFunctionInst *inst, SILBuilder &builder, SILLocation loc,
       DifferentiationInvoker invoker);
 
 public:
+  /// Construct an DifferentiationTransformer for the given module.
+  explicit DifferentiationTransformer(SILModuleTransform &transform)
+    : transform(transform), context(transform) {}
+
+  ADContext& getContext() { return context; }
```

Comment:
The current style is to put a space before `&`, not after. @bgogul: could you please address in this a follow-up?
```cpp
  ADContext &getContext() { return context; }
```